codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt

```diff
@@ -858,12 +862,22 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
                                 // `<_>::from()` is necessary to convert the `&str` into:
                                 //     * the Rust enum in case the `string` shape has the `enum` trait; or
                                 //     * `String` in case it doesn't.
-                                rustTemplate(
-                                    """
-                                    let v = <_>::from(#{PercentEncoding}::percent_decode_str(&v).decode_utf8()?.as_ref());
-                                    """.trimIndent(),
-                                    *codegenScope
-                                )
+                                if (memberShape.hasTrait<EnumTrait>()) {
+                                    rustTemplate(
+                                        """
+                                        let v = <#{memberShape}>::try_from(#{PercentEncoding}::percent_decode_str(&v).decode_utf8()?.as_ref())?;
+                                        """,
+                                        *codegenScope,
+                                        "memberShape" to symbolProvider.toSymbol(memberShape),
+                                    )
+                                } else {
+                                    rustTemplate(
+                                        """
+                                        let v = <_>::from(#{PercentEncoding}::percent_decode_str(&v).decode_utf8()?.as_ref());
+                                        """.trimIndent(),
```

Comment:
nit: you'll sometimes see these `trimIndent` sprinkled over the codebase. They are added by IntelliJ when hitting Enter inside a string block. We decided some time ago that we would not insert these anymore when generating Rust code, since we run `cargo fmt` after generation, so they just add noise. If you touch a part of the codebase where you see one of these, feel free to remove it.

I guess we could also just grep them out in one commit and forget about them, but I haven't gotten around to doing that just yet. Feel free to take it on if you want.